pkg/controller/persistentvolume/state.go

```diff
@@ -0,0 +1,225 @@
+)
+
+type persistentVolumeStateContext struct {
+	volume        *api.PersistentVolume
+	claim         *api.PersistentVolumeClaim
+	claimNotFound bool
+	client        binderClient
+	volumeIndex   *persistentVolumeOrderedIndex
+	locks         keymutex.KeyMutex
+	states        []persistentVolumeState
+}
+
+type persistentVolumeState interface {
+	isCurrentState(context *persistentVolumeStateContext) bool
+	processState(context *persistentVolumeStateContext) error
+}
+
+// newPersistentVolumeStateContext is a factory function that creates persistentVolumeStateContext objects with
+// the correct internal state object set according to the phase of the PersistentVolume.
+// This factor also attempts to find any claim bound to the volume via the volume's ClaimRef
+func newPersistentVolumeStateContext(pv *api.PersistentVolume, client binderClient, volumeIndex *persistentVolumeOrderedIndex, locks keymutex.KeyMutex) (*persistentVolumeStateContext, error) {
+	ctx := &persistentVolumeStateContext{
+		client:      client,
+		volume:      pv,
+		volumeIndex: volumeIndex,
+		locks:       locks,
+		states: []persistentVolumeState{
+			&persistentVolumePendingState{},
+			&persistentVolumeBoundState{},
+			&persistentVolumeReleasedState{},
+		},
+	}
+	if err := ctx.fetchClaim(); err != nil {
+		return nil, err
+	}
+	return ctx, nil
+}
+
+// fetchClaim will find the claim bound to the PV, if any, and set it on the persistentVolumeContext object.
+// any error that is not 404 will be returned.  Missing claims (404s) are normal and not fatal to the PV's state.
+// Claims that 404 are important to track, because a deleted claim is meaningful to a volume bound to it.
+func (ctx *persistentVolumeStateContext) fetchClaim() error {
+	if ctx.volume != nil && ctx.volume.Spec.ClaimRef != nil {
+		claim, err := ctx.client.GetPersistentVolumeClaim(ctx.volume.Spec.ClaimRef.Namespace, ctx.volume.Spec.ClaimRef.Name)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		if err != nil && errors.IsNotFound(err) {
+			ctx.claimNotFound = true
+			return nil
+		}
+		ctx.claim = claim
+		ctx.claimNotFound = false
+	}
+	return nil
+}
+
+//makeCloneForUpdate clones a volume that is then used to update the API.
+// a clone is used because modifying the PV itself means modifying the volumes held by the controller (it's a pointer)
+// as well as having to rollback those changes in the event of API failure.
+func makeCloneForUpdate(pv *api.PersistentVolume) (*api.PersistentVolume, error) {
+	clone, err := conversion.NewCloner().DeepCopy(pv)
+	if err != nil {
+		return nil, fmt.Errorf("Error cloning pv: %v", err)
+	}
+	volumeClone, ok := clone.(*api.PersistentVolume)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
+	}
+	return volumeClone, nil
+}
+
+func (ctx *persistentVolumeStateContext) updatePersistentVolumeStatus(phase api.PersistentVolumePhase) error {
+	if ctx.volume.Status.Phase == phase {
+		return nil
+	}
```

Comment:
The controllers are all designed to build state by observation, specifically so that we can rebuild after a crash.

The state objects proposed look at the existing state and put the PV/C into the correct phase.